Source/DafnyCore/Verifier/Translator.cs

```diff
@@ -125,6 +126,55 @@ public class TranslatorFlags {
     readonly Dictionary<string, Bpl.Constant> fieldConstants = new Dictionary<string, Constant>();
     readonly Dictionary<string, Bpl.Constant> tytagConstants = new Dictionary<string, Constant>();
 
+    public class ProofDependencyManager {
+      // proof dependency tracking state
+      public Dictionary<string, ProofDependency> ProofDependenciesById { get; } = new();
+      private UInt64 proofDependencyIdCount = 0;
+
+      public string GetProofDependencyId(ProofDependency dep) {
+        var idString = $"id{proofDependencyIdCount}";
+        ProofDependenciesById[idString] = dep;
+        proofDependencyIdCount++;
+        return idString;
+      }
+
+      public void AddProofDependencyId(ICarriesAttributes boogieNode, IToken tok, ProofDependency dep) {
+        var idString = GetProofDependencyId(dep);
+        boogieNode.Attributes =
+          new QKeyValue(tok, "id", new List<object>() { idString }, boogieNode.Attributes);
+      }
+
+      // Get the full ProofDependency indicated by a compound ID string.
+      public ProofDependency GetFullIdDependency(string idString) {
+        var parts = idString.Split('$');
```

Comment:
I think string parsing should be reserved for working with the outside world, not for in-memory information processing.